packages/react-ui/src/widgets/time-series-widget-ui/hooks/TimeSeriesContext.js

```diff
@@ -0,0 +1,147 @@
+  setTimeframe: (value) => {},
+  onTimeframeUpdate: (value) => {},
+  togglePlay: () => {},
+  stop: () => {}
+});
+
+export function useTimeSeriesContext() {
+  return useContext(TimeSeriesContext);
+}
+
+export function TimeSeriesProvider({
+  children,
+  animationStep,
+  isPlaying,
+  isPaused,
+  onPlay,
+  onPause,
+  onStop,
+  timelinePosition,
+  onTimelineUpdate,
+  timeframe,
+  onTimeframeUpdate
+}) {
+  const [_isPlaying, setIsPlaying] = useState(isPlaying);
+  const [_isPaused, setIsPaused] = useState(isPaused);
+  const [_timelinePosition, setTimelinePosition] = useState(0);
+  const [_timeframe, setTimeframe] = useState([]);
+
+  useEffect(() => {
+    if (_timeframe.length === 2) {
+      onTimeframeUpdate(_timeframe.sort());
+    }
+    // Only executed when timeframe changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [_timeframe]);
+
+  useEffect(() => {
+    if (!_timeframe.length && (_isPlaying || _isPaused)) {
+      onTimelineUpdate(_timelinePosition);
+    }
+    // Only executed when timelinePosition changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [_timelinePosition, _isPlaying, _isPaused, _timeframe]);
+
+  const togglePlay = useCallback(() => {
+    if (_isPlaying) {
+      if (onPause) onPause();
+      setIsPaused(true);
+    }
+    if (!_isPlaying) {
+      if (onPlay) onPlay();
+      setIsPaused(false);
+    }
+    setIsPlaying((oldState) => !oldState);
+  }, [_isPlaying, onPause, onPlay]);
+
+  const stopWrapper = useCallback(() => {
+    setIsPlaying(false);
+    setIsPaused(false);
+    setTimelinePosition(0);
+    setTimeframe([]);
+  }, []);
+
+  const stop = useCallback(() => {
+    stopWrapper();
+    // Wait a clock cycle to avoid condition race
+    // meanwhile the states is setted above
+    setTimeout(onStop || (() => {}));
+  }, [onStop, stopWrapper]);
+
+  // These four useEffects are used to allow
+  // manage time series state outside the widget
+  useEffect(() => {
+    if (isPlaying !== _isPlaying) {
+      setIsPlaying(isPlaying);
+    }
+    // Only when the state out of the context changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isPlaying]);
+
+  useEffect(() => {
+    if (isPaused !== _isPaused) {
+      setIsPaused(isPaused);
+    }
+    // Only when the state out of the context changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isPaused]);
+
+  useEffect(() => {
+    if (timelinePosition !== _timelinePosition) {
+      setTimelinePosition(timelinePosition);
+    }
+    // Only when the state out of the context changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [timelinePosition]);
+
+  useEffect(() => {
+    if (timeframe[0] !== _timeframe[0] && timeframe[1] !== _timeframe[1]) {
+      setTimeframe(timeframe);
+    }
+    // Only when the state out of the context changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [...timeframe]);
```

Comment:
This useEffect condition is correct? It always returns a new object so what is the sense of using a useEffect?

Reply:
Yes, it's correct. Array of dependencies is always a new object, but it isn't what useEffect checks, it checks every element inside that array of dependencies. That's why I spread the timeframe array, because I don't want to check new reference by timeframe, I want to check inner values because the timeframe array can change but its inner values no.

I don't want to check this: [[1,1]], I want to check: [1,1].

I think it's a minor improvement that might not be absolutely necessary, if it's annoying for you, I can delete it.